crates/ball_filter/src/hypothesis_merger.rs

```diff
@@ -0,0 +1,58 @@
+use crate::{hypothesis::BallHypothesis, BallFilter};
+
+pub struct RemovedHypotheses(pub(crate) Vec<BallHypothesis>);
+
+impl RemovedHypotheses {
+    pub fn inner(self) -> Vec<BallHypothesis> {
+        self.0
+    }
+}
+
+pub struct ValidHypotheses(Vec<BallHypothesis>);
+
+pub trait HypothesisMerger {
+    fn partition(
+        &mut self,
+        is_valid: impl Fn(&BallHypothesis) -> bool,
+    ) -> (ValidHypotheses, RemovedHypotheses);
+    fn merge(
+        &mut self,
+        valid_hypotheses: ValidHypotheses,
+        merge_criterion: impl Fn(&BallHypothesis, &BallHypothesis) -> bool,
+    );
+}
+
+impl HypothesisMerger for BallFilter {
+    fn partition(
+        &mut self,
+        is_valid: impl Fn(&BallHypothesis) -> bool,
+    ) -> (ValidHypotheses, RemovedHypotheses) {
+        let (valid, removed) = self.hypotheses.drain(..).partition(is_valid);
+        (ValidHypotheses(valid), RemovedHypotheses(removed))
+    }
```

Comment:
the `hypotheses` of the ball filter are private. Also I think the wrapper types are useful, since it is not possible to use the incorrect hypotheses in the ball filter api

Reply:
removed